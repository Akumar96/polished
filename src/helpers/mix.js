// @flow

import rgba from './rgba'
import parseColorString from '../internalHelpers/_parseColorString'

/**
 * Mixes two colors together.
 *
 * @example
 * // Styles as object usage
 * const styles = {
 *   background: mix('#f00', '#00f')
 *   background: mix(0.25, '#f00', '#00f')
 *   background: mix('rgba(255, 0, 0, 0.5)', '#00f')
 * }
 *
 * // styled-components usage
 * const div = styled.div`
 *   background: ${mix('#f00', '#00f')};
 *   background: ${mix(0.25, '#f00', '#00f')};
 *   background: ${mix('rgba(255, 0, 0, 0.5)', '#00f')};
 * `
 *
 * // CSS in JS Output
 *
 * element {
 *   background: "#7f007f";
 *   background: "#3f00bf";
 *   background: "rgba(63, 0, 191, 0.75)";
 * }
 */
function mix(colorOrWeight: string | number, color: string, otherColor?: string): string {
  const weight = typeof colorOrWeight === 'number' ? colorOrWeight : 0.5
  const colorString1 = typeof colorOrWeight === 'number' ? color : colorOrWeight
  const colorString2 = typeof colorOrWeight === 'number' ? otherColor : color

  const parsedColor1 = parseColorString(colorString1)
  const color1 = {
    ...parsedColor1,
    alpha: typeof parsedColor1.alpha === 'number' ? parsedColor1.alpha : 1,
  }

  const parsedColor2 = parseColorString(colorString2)
  const color2 = {
    ...parsedColor2,
    alpha: typeof parsedColor2.alpha === 'number' ? parsedColor2.alpha : 1,
  }

  // The formular is copied by the original Sass implementation:
  // http://sass-lang.com/documentation/Sass/Script/Functions.html#mix-instance_method
  const alphaDelta = color1.alpha - color2.alpha
  const x = (weight * 2) - 1
  const y = x * alphaDelta === -1 ? x : (x + alphaDelta)
  const z = (1 + (x * alphaDelta))
  const weight1 = ((y / z) + 1) / 2.0
  const weight2 = 1 - weight1

  const mixedColor = {
    red: Math.floor((color1.red * weight1) + (color2.red * weight2)),
    green: Math.floor((color1.green * weight1) + (color2.green * weight2)),
    blue: Math.floor((color1.blue * weight1) + (color2.blue * weight2)),
    alpha: color1.alpha + ((color2.alpha - color1.alpha) * (weight / 1.0)),
  }

  return rgba(mixedColor)
}

export default mix
