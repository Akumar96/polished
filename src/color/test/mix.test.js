// @flow
import mix from '../mix'

describe('mix', () => {
  it('should mix two colors with by a weight of 25%', () => {
    expect(mix(0.25, '#f00', '#00f')).toMatchSnapshot()
  })

  it('should mix two colors with an 8-digit hex color', () => {
    expect(mix(0.5, '#FF00007F', '#00f')).toMatchSnapshot()
  })

  it('should mix two colors with a 4-digit hex color', () => {
    expect(mix(0.5, '#FF00007F', '#0f08')).toMatchSnapshot()
  })

  it('should mix two colors with opacity lower than 1', () => {
    expect(mix(0.5, 'rgba(255, 0, 0, 0.5)', '#00f')).toMatchSnapshot()
  })

  it('should mix two colors when weight is a string', () => {
    expect(mix('0.5', 'rgba(255, 0, 0, 0.5)', '#00f')).toMatchSnapshot()
  })

  it('should return otherColor when passed transparent color', () => {
    expect(mix('0.5', 'transparent', '#00f')).toMatchSnapshot()
  })

  it('should return color when passed transparent otherColor', () => {
    expect(mix('0.5', 'rgba(255, 0, 0, 0.5)', 'transparent')).toMatchSnapshot()
  })

  it('should return transparent when passed transparent for both colors', () => {
    expect(mix('0.5', 'transparent', 'transparent')).toMatchSnapshot()
  })
})
