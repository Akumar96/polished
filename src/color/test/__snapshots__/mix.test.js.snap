// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mix should mix two colors when weight is a string 1`] = `"rgba(63,0,191,0.75)"`;

exports[`mix should mix two colors with a 4-digit hex color 1`] = `"rgba(123,131,0,0.515)"`;

exports[`mix should mix two colors with an 8-digit hex color 1`] = `"rgba(63,0,191,0.75)"`;

exports[`mix should mix two colors with by a weight of 25% 1`] = `"#3f00bf"`;

exports[`mix should mix two colors with opacity lower than 1 1`] = `"rgba(63,0,191,0.75)"`;

exports[`mix should return color when passed transparent otherColor 1`] = `"rgba(255, 0, 0, 0.5)"`;

exports[`mix should return otherColor when passed transparent color 1`] = `"#00f"`;

exports[`mix should return transparent when passed transparent for both colors 1`] = `"transparent"`;
